### ISSUE TITLE
server-side polyfill

### ISSUE THREAD

**Issue by mhrisse** [igrigorik/http-client-hints#50]
Hello Ilya,

I am really happy you are approaching this topic so constructively.

While proposal process and implementation is undertaken, I was wondering if one could use the device detection engine of various CDN vendors (Akamai, Edgecast, CloudFront, .. ) or your own caching proxy to create some sort of serve-side polyfill so we can start implementing this using the new HTTP headers now? So the idea is to have a simple UA-string to Client Hints "function" in the CDN layer.

### Example

Assuming a iPhone 6 with Safari Mobile is accessing a webpage URL (let's say HTML for simplicity) and does *not* send client hints itself. Would it be meaningful to let the device detection engine in the CDN translate the UA string to client hint headers, if no client hints are present. Of course I am mostly concerned about mobile devices, given they are the weakest clients.

#### Original request
```
Mozilla/5.0 (iPhone; CPU iPhone OS 7_0 like Mac OS X; en-us) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53
```

could result for example in adding the following client hint headers to the request. 

#### Modified request
```
DPR: 2.0
Width: 160
Viewport-Width: 320
```

Then, the webserver could deliver optimized HTML to the client, as described in the proposal to optimize the critical render path. Probably not perfect, but surely better than speedtest detection and other measures that require javascript execution and roundtrips?

Of course, the CDN must be configured to cache this URL + header combination from now on to preserve cache efficiency. 

Is this totally a bad idea or do you have any thoughts on this?

PS: I would deliberately exclude ```Downlink```, I guess that is not really possible without Javascript being executed on the client?

**Comment by yoavweiss** [comment]
Answering this for Ilya, since he's OOO AFAIK :)

For `DPR` and `Viewport-Width` you could do that to some degree of accuracy on mobile devices. You won't be able to do the same for desktops or devices where the viewport can occupy only part of the available screen.
`Width` OTOH is not something you can get on the server side without maintaining a *lot* of state.
Also, you be missing client-side support for `Content-DPR` which would mean you'd have to be careful about having all image dimensions well defined in CSS or HTML, or layout would break.

Regarding your question if this is a good idea, I guess my question would be "What's the use case?". Why would you "polyfill" CH on the server side instead of just acting upon that (incomplete) info right there?

**Comment by ejthurgo** [comment]
Just my humble 2p. There are services out there which do this already, but ultimately they are flawed as Yoav says. Not to mention the maintenance overhead of translating the myriad of UA strings to usable data.
I'm not sure how you would determine 'width' from the UA string, you could only reliably get device width, which is also not equivalent to viewport width as it changes based on device orientation. In terms of images, viewport width carries little value because images rarely occupy the full width, resulting in superfluous data download. Outside of images, viewport width definitely has value though.

I think the drive for Client Hints is to modify browsers to send 'real' data rather than servers having to guess from a UA String.

**Comment by mhrisse** [comment]
Thanks for your thoughts.

I am aware this is only a non-perfect stopgap solution until client hints are there from the browser.

Afaik believe some of the CDN device detection services in the wild who give you properties based on UA-string are based on the WURFL database:

http://wurfl.sourceforge.net/help_doc.php#html_ui
http://wurfl.sourceforge.net/help_doc.php#display

Pseudo-code:

``` 
// Viewport-Width
Viewport-Width: max(resolution_width, resolution_height) // maybe viewport_width

// DPR:
if(device_os != "iOS") { DPR: max(physical_screen_width / resolution_width) }
```

Looks like determining the DPR for Android is potentially possible, but not for iOS devices: https://www.scientiamobile.com/forum/viewtopic.php?f=16&t=286 - needless to say this is no good :)

@yoavweiss The use-case is to optimize the critical render path (especially on mobile) and minimize asset loading while keeping a heavy traffic'ed site CDN-cacheable. I guess my question basically is if it better to implement the server-behaviour against this new standard later send by the browser, rather than implementing my own little API.

Thank you for thinking this thru with me. :)

**Comment by yoavweiss** [comment]
> @yoavweiss The use-case is to optimize the critical render path (especially on mobile) and minimize asset loading while keeping a heavy traffic'ed site CDN-cacheable. I guess my question basically is if it better to implement the server-behaviour against this new standard later send by the browser, rather than implementing my own little API.

OK, I see what you're asking. I guess it can make sense if you want to have a single server-side layer that performs image manipulation, and add some hints (lower in the stack) on requests coming from non-supporting browsers. It all depends on your server architecture.

**Comment by igrigorik** [comment]
The primary goal of the entire CH exercise is to converge on a common set of header names and values (i.e. their format). As such, who and how they emit these hints is completely open... If you have an intermediary that is able to infer such things, then it's completely reasonable for it to emit these hints to a downstream server. Which is to say, I wouldn't even call it a "server polyfill" because I see this type of use case as core to the experience (e.g. you can make an argument that Service Worker is one such example). 

That said, there are limitations about what values you'll be able to infer downstream from the client: we know that UA sniffing is hard and mostly broken (but it works well enough for some folks); dynamic values (e.g. display width, downlink speed, etc) can't be inferred downstream, and so on.

In short, it's a good idea, and I hope that we'll see CDNs/proxies converge on using CH hints.

**Comment by igrigorik** [comment]
Closing, don't think there is anything actionable here from CH spec perspective.

**Comment by mhrisse** [comment]
Thank you for your thoughts and in-depth answers, guys!
